### PR TITLE
[ckpt] fix: skip reload_model_params for fsdp_dtensor checkpoints

### PR DESCRIPTION
**Pull request by denys-fridman**
MCore commit ab43d43f0 changed DistributedOptimizer._copy_model_params_to_main_params from a silent no-op to a NotImplementedError for Megatron-FSDP. This breaks FSDP DTensor checkpoint loading when load_optim=False with bf16/fp16, since the else branch calls optimizer.reload_model_params() unconditionally.

Skip the call for fsdp_dtensor format, matching the previous mcore behavior.

# What does this PR do ?

Add a one line overview of what this PR aims to accomplish.

# Changelog

- Add specific line by line info of high level changes in this PR.

# GitHub Actions CI

See the [CI section](https://github.com/NVIDIA-NeMo/Megatron-Bridge/blob/main/CONTRIBUTING.md#-running-github-ci)in the Contributing doc for how to trigger the CI. A Nvidia developer will need to approve and trigger the CI for external contributors.

# Before your PR is "Ready for review"

**Pre checks**:

- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA-NeMo/Megatron-Bridge/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?

If you haven't finished some of the above items you can still open "Draft" PR.

# Additional Information

- Related to # (issue)

## Summary by CodeRabbit

* **Bug Fixes**
  * Optimized checkpoint loading to prevent unnecessary parameter operations for specific checkpoint formats when mixed-precision training is enabled, improving training stability and efficiency.

### CODE REVIEW

**Review by cspades — CHANGES_REQUESTED** [review]
This will break checkpoint installation with Megatron-FSDP - the main weights will be loaded but compute weights & quantization will be skipped for the first global step.

Will recommend a solution ASAP.

File: src/megatron/bridge/training/checkpointing.py [hunk 1]
```diff
@@ -2163,7 +2163,7 @@ def _load_checkpoint_from_path(
             )
             raise e
     else:
-        if (cfg.model.fp16 or cfg.model.bf16) and optimizer is not None:
+        if (cfg.model.fp16 or cfg.model.bf16) and optimizer is not None and not skip_load_to_model_and_opt:
```
Comment (cspades) [inline src/megatron/bridge/training/checkpointing.py]:
```suggestion
        if (cfg.model.fp16 or cfg.model.bf16) and optimizer is not None and not cfg.ddp.use_megatron_fsdp:
```

File: src/megatron/bridge/training/setup.py [hunk 2]
```diff
@@ -279,7 +279,7 @@ def modelopt_pre_wrap_hook(model):
             model=model,
             optimizer=optimizer,
             opt_param_scheduler=scheduler,
-            skip_load_to_model_and_opt=cfg.dist.use_torch_fsdp2,
+            skip_load_to_model_and_opt=cfg.dist.use_torch_fsdp2 or cfg.dist.use_megatron_fsdp,
```
Comment (cspades) [inline src/megatron/bridge/training/setup.py]:
```suggestion
            skip_load_to_model_and_opt=cfg.dist.use_torch_fsdp2,
```

**Review by cspades — APPROVED** [review]

**Comment by cuichenx** [comment]
/ok to test https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/3425/commits/36a8595edd1f3e70ddf46c7c5b6823757b889719

**Comment by pruprakash** [comment]
## QA RCCA Analysis

### 1. Fix Reference
- **PR**: #3425 - [ckpt] fix: skip reload_model_params for fsdp_dtensor checkpoints
- **Issue**: PR is the fix itself (no linked issue)

### 2. Root Cause
MCore commit ab43d43f0 changed `DistributedOptimizer._copy_model_params_to_main_params` from a silent no-op to a `NotImplementedError` for Megatron-FSDP. This breaks FSDP DTensor checkpoint loading when `load_optim=False` with bf16/fp16, since the else branch calls `optimizer.reload_model_params()` unconditionally.

### 3. Trigger Configuration
- Loading fsdp_dtensor checkpoints with `load_optim=False`
- Using bf16 or fp16 precision
- Updated MCore version with the breaking change

### 4. Nature of the Bug
**Classification**: CODE BUG - Checkpoint loading failure due to MCore API change

### 5. Existing Test Coverage
**In Fix PR**: NO - No test files in PR

**In NMFW Tests**: Not specifically covering fsdp_dtensor checkpoint loading with load_optim=False

### 6. Coverage Assessment
| Test Type | Exists | Covers Bug |
|-----------|--------|------------|
| Fix PR unit tests | NO | N/A |
| NMFW regression tests | NO | N/A |

### 7. New Regression Test
**NOT ADDED** - This is a compatibility fix for MCore API change. The specific scenario (fsdp_dtensor + load_optim=False + bf16/fp16) is an edge case. FSDP functional tests in the repo provide general coverage.

### 8. Conclusion
**Verdict**: PARTIAL COVERAGE - No dedicated test for this specific edge case, but FSDP functional tests provide general checkpoint coverage.
